internal/scraper/scraper.go

```diff
@@ -220,74 +222,113 @@ func (s *Scraper) Run(ctx context.Context) {
 	// TODO(mem): keep count of the number of successive errors and
 	// collect logs if threshold is reached.
 
-	var sm checkStateMachine
+	var (
+		frequency = ms(s.check.Frequency)
+		offset    = ms(s.check.Offset)
+	)
 
-	// need to keep the most recently published payload for clean up
-	var payload *probeData
+	if offset == 0 {
+		offset = randDuration(min(frequency, maxPublishInterval))
+	}
 
-	scrape := func(ctx context.Context, t time.Time) {
-		s.scrapeCounter.Inc()
+	scrapeHandler := scrapeHandler{scraper: s}
 
-		var err error
-		payload, err = s.collectData(ctx, t)
+	tickWithOffset(
+		ctx,
+		s.stop,
+		scrapeHandler.scrape, scrapeHandler.republish, scrapeHandler.cleanup,
+		frequency, offset,
+		maxPublishInterval, minPublishGap,
+	)
 
-		switch {
-		case errors.Is(err, errCheckFailed):
-			s.errorCounter.WithLabelValues("check").Inc()
-			sm.fail(func() {
-				s.logger.Info().Msg("check entered FAIL state")
-			})
-
-		case err != nil:
-			s.errorCounter.WithLabelValues("collector").Inc()
-			s.logger.Error().Err(err).Msg("error collecting data")
-			return
+	s.cancel()
 
-		default:
-			sm.pass(func() {
-				s.logger.Info().Msg("check entered PASS state")
-			})
-		}
+	s.logger.Info().Msg("scraper stopped")
+}
 
-		if payload != nil {
-			s.publisher.Publish(payload)
-		}
+type scrapeHandler struct {
+	scraper *Scraper
+	payload *probeData // needed to keep the most recently published payload for republication and clean up
+	sm      checkStateMachine
+}
+
+func (h *scrapeHandler) scrape(ctx context.Context, t time.Time) {
+	h.scraper.scrapeCounter.Inc()
+
+	var err error
+	h.payload, err = h.scraper.collectData(ctx, t)
+
+	switch {
+	case errors.Is(err, errCheckFailed):
+		h.scraper.errorCounter.WithLabelValues("check").Inc()
+		h.sm.fail(func() {
+			h.scraper.logger.Info().Msg("check entered FAIL state")
+		})
+
+	case err != nil:
+		h.scraper.errorCounter.WithLabelValues("collector").Inc()
+		h.scraper.logger.Error().Err(err).Msg("error collecting data")
+		return
+
+	default:
+		h.sm.pass(func() {
+			h.scraper.logger.Info().Msg("check entered PASS state")
+		})
 	}
 
-	cleanup := func(ctx context.Context, t time.Time) {
-		if payload == nil {
-			return
-		}
+	if h.payload != nil {
+		h.scraper.publisher.Publish(h.payload)
+	}
+}
 
-		staleSample := prompb.Sample{
-			Timestamp: t.UnixNano()/1e6 + 1, // ms
-			Value:     staleMarker,
-		}
+func (h *scrapeHandler) republish(ctx context.Context, t time.Time) {
+	if h.payload == nil {
+		return
+	}
 
-		for i := range payload.ts {
-			ts := &payload.ts[i]
-			for j := range ts.Samples {
-				ts.Samples[j] = staleSample
-			}
+	h.payload.streams = nil // do not republish logs
+
+	now := t.UnixMilli()
+	for i := range h.payload.ts {
+		ts := &h.payload.ts[i]
+		for j := range ts.Samples {
+			ts.Samples[j].Timestamp = now
```

Comment:
I might be misunderstanding something here, but, if I'm understanding this correctly, once we want to republish the previous published data, we iterate through all the previously collected time series, and for each sample in the time series we set the current time. That means that, if some of these time series collected for a check have more than one sample, when republishing all of them will have the same timestamp.

Is that correct? In such case, can that produce weird representations of the data? As I assume that the previously generated time series, probably have different timestamp values per sample..?

I wonder if we should define the sample timestamps based on the current time and also on the delta time between the samples for the "original payload", so basically maintaining the time difference between samples. Although for that we would have to, either define timestamps that are "in the future", or do it backwards, so the "oldest" sample for each time series will get the _timestamp=now_ and then we would recalculate the rest subtracting their respective deltas from _now_ back in time.

Let me know if I misunderstood something, or if I did not explain myself properly.

Reply:
No, all the data in payload has the same timestamp (#L259, #L501, #L734, #L793). Queuing older data (or data with different timestamps) happens in the publisher, not in the scraper. The data collected in the scraper corresponds to a single point in time.

What you are saying is possible, but it doesn't happen here.

Reply:
Got it! Thanks for the explanation.